### PR TITLE
Random and Reverse Iterators

### PR DESCRIPTION
**Pull request by adsharma**
Two simple iterators:

* Random: uses reservoir sampling
* Reverse: uses O(n) storage to reverse in memory. Could be optimized.

### CODE REVIEW

File: src/RandomIterator.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#include "iterlib/RandomIterator.h"
+
+#include <folly/Random.h>
+
+namespace iterlib {
+
+bool RandomIterator::doNext() {
```
Comment (manukranthk) [inline src/RandomIterator.cpp]: This can be optimized to make the get random samples lazy load.

Comment (adsharma) [inline src/RandomIterator.cpp]: How do you propose we do that without knowing how many items are there in the inner iterator? Related discussion:

http://stackoverflow.com/questions/24172693/iterative-or-lazy-reservoir-sampling

Comment (manukranthk) [inline src/RandomIterator.cpp]: I agree. Spoke without completely understanding the solution here.

File: include/iterlib/RandomIterator.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#include "iterlib/WrappedIterator.h"
+
+namespace iterlib {
+
+// get random results from Iterator
+class RandomIterator : public WrappedIterator {
+public:
+  RandomIterator(Iterator* iter, int count)
+    : WrappedIterator(iter)
+    , count_(count)
```
Comment (manukranthk) [inline include/iterlib/RandomIterator.h]: Can you add this in the comments : 
https://en.wikipedia.org/wiki/Reservoir_sampling
Also, mention what happens if the inner iterator has lesser elements than count?

**Pull request merged**